index/scorch/snapshot_segment.go

```diff
@@ -215,9 +215,32 @@ func (cfd *cachedFieldDocs) prepareFields(field string, ss *SegmentSnapshot) {
 }
 
 type cachedDocs struct {
-	m     sync.Mutex                  // As the cache is asynchronously prepared, need a lock
-	cache map[string]*cachedFieldDocs // Keyed by field
-	size  uint64
+	m       sync.Mutex                  // As the cache is asynchronously prepared, need a lock
+	cache   map[string]*cachedFieldDocs // Keyed by field
+	size    uint64
+	pfields []string
+}
+
+func (c *cachedDocs) getCachedFields() []string {
+	rv := make([]string, len(c.cache))
+	c.m.Lock()
+	for k, _ := range c.cache {
+		rv = append(rv, k)
+	}
+	c.m.Unlock()
+	return rv
+}
+
+func (c *cachedDocs) getPersistedFields() []string {
+	c.m.Lock()
```

Comment:
the lock area on these methods is perhaps too small, where a caller can get the persisted or cached fields, does some work, but in the meantime another thread can sneak in an change the persisted / cached fields?

Reply:
so, i agree in principle, however here the thinking is this:

the persisted fields is actually immutable, it's just that someone has to do the work to load it and figure out which fields it is.  yes, if the first time in, multiple threads race to do it, and they both do, they always come to the same conclusion.